### ISSUE TITLE
jit silently miscompiles `alpha * (A @ B + bias)` on sm_120: `alpha` is applied to the matmul product but not to the bias

### ISSUE THREAD

**Issue by adam-hartshorne** [jax-ml/jax#39336]
## Summary

On an RTX 5090 (Blackwell, sm_120) with jax/jaxlib 0.10.2, a scalar compile-time constant
multiplying a linear layer is fused into the GEMM epilogue, but the scale is applied **only to
the matmul product** — the bias addend is left unscaled:

```
computed:  alpha * (A @ B) + bias
correct:   alpha * (A @ B) + alpha * bias
```

The result is wrong by exactly `(1 - alpha) * bias`: a constant offset per output channel,
identical for every row of the batch. Eager, the CPU backend, and `--xla_gpu_autotune_level<=3`
are all correct; GPU jit at the default autotune level is not. The wrong answer is
bit-deterministic across repeated calls, so this is not non-determinism or a tolerance issue.

**This is a large error, not a precision one.** At `alpha=0.1` — an unremarkable choice for
scaling down a layer's output — the compiled program's result has a **199% RMS relative error**:
it is twice as wrong as the quantity is big. It is also completely silent: no warning, no NaN,
no non-determinism, and eager mode reproduces the correct answer, so nothing in a normal
workflow flags it.

## Reproducer

```python
import jax, jax.numpy as jnp, jax.random as jr, numpy as np

k = jr.split(jr.PRNGKey(0), 6)
P = (jr.normal(k[0], (32, 15)) * 0.3, jr.normal(k[1], (32,)) * 0.1,
     jr.normal(k[2], (32, 32)) * 0.3, jr.normal(k[3], (32,)) * 0.1,
     jr.normal(k[4], (3, 32)) * 0.1,  jr.normal(k[5], (3,)) * 0.1)
t = jnp.linspace(0., 1., 100)

def net(p, tau):
    W1, b1, W2, b2, W3, b3 = p
    h = jax.nn.gelu(W1 @ jnp.full((15,), tau) + b1)
    h = jax.nn.gelu(W2 @ h + b2)
    return W3 @ h + b3            # <- final GEMM + bias

for alpha in (0.1, 0.5, 2.0, 1.0):
    f = lambda p, tt: alpha * jax.vmap(lambda x: net(p, x))(tt)
    D = np.asarray(jax.jit(f)(P, t)) - np.asarray(f(P, t))
    print(f"alpha={alpha}  offset {np.round(D.mean(0), 7)}  "
          f"predicted (1-a)*bias {np.round((1 - alpha) * np.asarray(P[5]), 7)}")
```

Expected: every offset is zero.

### How large the error is

This is not a precision or tolerance question. The smaller `alpha` is, the more the unscaled
bias dominates, and scaling a layer's output down by ~0.1 is an ordinary thing to do:

| alpha | max abs of correct output | max abs error | error / max abs output | **RMS relative error** |
|---|---|---|---|---|
| **0.1** | 9.3080e-02 | 1.1419e-01 | **122.7%** | **199.1%** |
| 0.5 | 4.6540e-01 | 6.3438e-02 | 13.6% | 22.1% |
| 2.0 | 1.8616e+00 | 1.2688e-01 | 6.8% | 11.1% |
| 1.0 | 9.3080e-01 | 0.0000e+00 | 0.0% | 0.0% |

At `alpha=0.1` the compiled program's output is **twice as wrong as it is large**.

### Why it is the bias

The `jit - eager` difference is a constant per output channel — the same value in every one of
the 100 rows (per-row spread ~1e-6) — and it equals `(1 - alpha) * b3` to float32 precision.
The last column below is the *agreement between the observed error and that prediction*, i.e.
how exactly the hypothesis holds; it is not the size of the error:

| alpha | observed `jit - eager` (mean over rows) | predicted `(1-alpha)*b3` | \|observed − predicted\| |
|---|---|---|---|
| 0.1 | `[ 0.0573640  0.0287633  0.1141889]` | `[ 0.0573639  0.0287632  0.1141888]` | 7.45e-08 |
| 0.5 | `[ 0.0318688  0.0159795  0.0634382]` | `[ 0.0318689  0.0159796  0.0634382]` | 8.20e-08 |
| 2.0 | `[-0.0637382 -0.0319592 -0.1268768]` | `[-0.0637377 -0.0319592 -0.1268764]` | 4.62e-07 |
| 1.0 | `[-2e-07  -0e+00  -1e-07]`            | `[0. 0. 0.]`                          | 2.01e-07 |

## Conditions

- **Autotune level.** `--xla_gpu_autotune_level` 0, 1, 2 and 3 all give exactly 0 error; levels
  4 (the default) and 5 give 1.142e-01.
- **Backend.** CPU gives 0 error. Eager gives 0 error in every configuration.
- **Weights must be runtime arguments.** If `P` is captured as a closure constant, XLA
  constant-folds the expression and the error drops to ~2e-6 (gone).
- **Depends only on the bias.** A structurally different network sharing only the final layer
  reproduces a **bit-identical** 1.142e-01 error.

**Hypothesis:** an autotuned GEMM epilogue (cuBLASLt `alpha`, or a Triton epilogue) absorbs the
scalar multiply and applies it to the product before the bias is added rather than to the sum.
Not confirmed against XLA's source — the above is what the measurements pin down.

## Possibly related

[#27228](https://github.com/jax-ml/jax/issues/27228) — "Loss of precision in XLA backend with
vmap and scan" (open). Same ingredients: `vmap`, `gemm_fusion_autotuner.cc`, wrong results
rather than rounding (its buffer comparator reports `8.18504, expected 7.16491`), and
`--xla_gpu_autotune_level=0` makes it go away. Reported on RTX 3090 and A100, so if it is the
same underlying fault it is **not** Blackwell-specific and this report adds the mechanism
(`(1-alpha)*bias`) and a dependency-free reproducer.

## Workarounds

1. `XLA_FLAGS='--xla_gpu_autotune_level=3'` — global and correct. Over 15 epochs of a real
   training script: 90.0 s vs 88.8 s at level 4, ~1.3% slower. Level 0 also works, costs ~31%.
2. `jax.lax.optimization_barrier` between the linear layer and the multiply — works, but is
   site-specific and the docs note it has no derivative or batching rules.
3. Restructure so no scalar multiplies a bias-add, e.g. fold the constant in as
   `(alpha*W) @ h + alpha*b`. Verified to make eager and jit agree.

Flags that did **not** help: `--xla_gpu_enable_triton_gemm=false`,
`--xla_gpu_enable_latency_hiding_scheduler=false`, `--xla_disable_hlo_passes=fusion`,
`--xla_gpu_enable_command_buffer=`. Setting `jax.default_matmul_precision` to `highest`,
`float32` or `bfloat16` changed nothing.

## Impact

In a real training run the affected expression was `1e-1 * Linear(h)` in a network head. The
offset made the reported loss 15x optimistic (3.76e-4 reported vs 5.77e-3 actual) and corrupted
the gradients, since training differentiated the same compiled program. The severity point is
that jit and eager disagree by more than the quantity's own magnitude, deterministically and
silently.

### System info (python version, jaxlib version, accelerator, etc.)

jax     0.10.2
jaxlib  0.10.2
device  CudaDevice(id=0) — NVIDIA GeForce RTX 5090 (Blackwell, sm_120)
driver  595.71.05
OS      Linux 7.0.0-28-generic
Python  3.12.12

**Comment by Nebulised** [comment]
Hi
Same specs(GPU, driver version etc)

I can reproduce the problem using the below script. 

I observed the problem go away when running in CPU only mode, setting the autotune level to less than 4 or when running in 64 bit mode.

I tried reproducing the problem without the vmap but was unable to. Weirdly removing the outside lambda (as I have done for the f2 lambda func) made the problem go away. Also changing the shape of b from (100,15) -> (100, 15,1) and the bias from (32,) -> (32,1) at the same time also made the problem go away 

Hope this is of some help

Cheers

```python3 
import os
os.environ["XLA_FLAGS"] = "--xla_gpu_autotune_level=4"

import jax
# Enabling X64 the problem goes away
# jax.config.update("jax_enable_x64", True)


# Enabling CPU the problem goes away
# jax.config.update("jax_platform_name", "cpu")
import jax.numpy as jnp

a = jax.random.normal(key=jax.random.key(42),
                      shape=(32,15))

b = jax.random.normal(key=jax.random.key(426),
                      shape=(100, 15)) 

bias = jax.random.normal(key=jax.random.key(115),
                         shape=(32,)) * 100 # CHANGING THIS FROM (32,) to (32,1) the discrepancy no longer occurs 



def func(_alpha,_a, _bias, _b):
    return _alpha * ((_a @ _b) + _bias)


alpha = 0.1


f = lambda _a_arg,_bias_arg, tt: jax.vmap(lambda x: func(alpha, _a_arg,_bias_arg, x))(tt)
f2 = jax.vmap(lambda x : func(alpha, a, bias, x)) # Outside lambda removed 


jitted_f_output = jax.jit(f)(a, bias,b)
non_jitted_f_output = f(a, bias, b)

jitted_f2_output = jax.jit(f2)(b)
non_jitted_f2_output = f2(b)


print(f"F Function Jit vs Non Jit | All close : {jnp.allclose(jitted_f_output, non_jitted_f_output)} | Average Abs Discrepancy : {jnp.mean(jnp.abs(jitted_f_output - non_jitted_f_output))}")
print(f"F2 Function Jit vs Non Jit | All close : {jnp.allclose(jitted_f2_output, non_jitted_f2_output)} | Average Abs Discrepancy : {jnp.mean(jnp.abs(jitted_f2_output - non_jitted_f2_output))}")


assert jnp.allclose(f2(b), f(a,bias, b))

```

Output of running this

```terminal
F Function Jit vs Non Jit | All close : False | Average Abs Discrepancy : 69.18453216552734
F2 Function Jit vs Non Jit | All close : True | Average Abs Discrepancy : 0.0
```

**Comment by jonkhler** [comment]
Independent confirmation, plus a manifestation that makes this bug particularly nasty for iterative solvers.

**Environment:** NVIDIA L4 (Ada, **sm_89**), jax/jaxlib **0.11.0**, driver 580.159.03 / CUDA 13.0, Linux x86_64, Python 3.12. So: not Blackwell-specific, and still present in 0.11.0.

**The scalar doesn't need to be an explicit `alpha *`.** A step-size division in an Euler integrator — `x + (h @ W + b) / num_steps` — triggers the same epilogue miscompilation when the loop is compiled via `lax.while_loop` / `fori_loop`. Two extra twists compared to the original report:

1. **Per-program autotuning makes it inconsistent within one workflow.** `jax.jit` of a *single* step compiled correctly for us (matches eager to ~4e-6); only the `while_loop`-body program picked the bad epilogue. In our case that meant training (jitted loss, no loop) was numerically fine while the sampling loop of the same network was silently wrong — the discrepancy is invisible to any train-time check.
2. **The error integrates.** Each iteration adds a constant `(1 − 1/n)·b`, so an `n`-step integrator drifts by ≈ `n·(1 − 1/n)·b ≈ n·b`. With `n = 64` and `b = [0.3, −0.2]`: predicted drift `[18.90, −12.60]`, measured `[18.83, −12.67]` (the small gap is the dynamics' feedback). We first saw this as a biased conditional distribution from a flow-matching sampler; the bias survived weeks of model-side debugging because everything about training looked correct.

Also matches the shape-dependence angle: batch ≥ ~8192 reproduces for us, batch 5000 does not (presumably kernel selection by shape).

**Training-free 20-line repro** (deterministic, `b ≠ 0` is the only requirement — zero-initialised biases hide the bug, which is why fresh networks don't reproduce it):

```python
import jax
import jax.numpy as jnp

BATCH = 20_000
w1 = 0.05 * jax.random.normal(jax.random.key(0), (2, 128))
w2 = 0.05 * jax.random.normal(jax.random.key(1), (128, 2))
bias = jnp.array([0.3, -0.2])

def step(carry):
    point, count = carry
    hidden = jnp.tanh(point @ w1)
    return point + (hidden @ w2 + bias) / 64.0, count + 1

point = jax.random.normal(jax.random.key(2), (BATCH, 2))
carry = (point, jnp.zeros((), jnp.int32))
for _ in range(64):
    carry = step(carry)
eager = carry[0]

looped, _ = jax.lax.while_loop(lambda c: c[1] < 64, step, (point, jnp.zeros((), jnp.int32)))
print("|while_loop - eager| max:", jnp.abs(looped - eager).max())   # 18.85 on L4, jax 0.11.0
print("mean drift:", looped.mean(0) - eager.mean(0))                # [ 18.83 -12.67]
print("predicted n*(1-1/n)*b:", 63.0 * bias)                        # [ 18.90 -12.60]
```

**Workarounds and non-workarounds, confirmed on sm_89 / 0.11.0** (all consistent with the original report):
- `--xla_gpu_autotune_level=3`: exact (1.2e-6); level 4 (default): 18.85; level 0: exact.
- CPU: exact (1.2e-7).
- `NVIDIA_TF32_OVERRIDE=0`, `JAX_DEFAULT_MATMUL_PRECISION=highest`, `--xla_gpu_enable_triton_gemm=false`, `--xla_gpu_enable_command_buffer=`, `--xla_gpu_enable_while_loop_double_buffering=false`: no effect.

**Comment by jonkhler** [comment]
Root cause located in XLA. Compiler-level evidence and the exact code path, from dumping HLO of the while_loop repro at autotune level 3 vs 4:

**Level 4 (wrong)** rewrites the second layer's Triton fusion into a cuBLASLt custom call:

```
%cublas-lt-matmul = custom-call(%hidden, %W2, %point, %bias),
    custom_call_target="__cublas$lt$matmul",
    gemm_backend_config = { alpha_real: 0.015625, beta: 1, epilogue: "BIAS", ... }
```

cuBLASLt computes `D = alpha*(A@B) + beta*C`, then the `BIAS` epilogue adds the bias vector **after** scaling. The source expression is `point + (A@B + bias)/64`, so the bias needed the `1/64` too. Level 3 keeps the whole thing as a `__triton_nested_gemm_fusion` (correct), which is why only levels >= 4 (where the cuBLAS fission candidate is tried) are affected.

**The bug** is in `GemmRewriter::HandleMultiply` ([gemm_rewriter.cc, `HandleMultiply`](https://github.com/openxla/xla/blob/2fa50644adea1d5a9bcea781bed98ad29141a4ef/xla/backends/gpu/transforms/gemm_rewriter.cc#L887)):

```cpp
if (config.beta() == 0.0 && existing_gemm->user_count() == 1) {
  complex128 prev_alpha = {config.alpha_real(), config.alpha_imag()};
  complex128 new_alpha = *alpha->literal().GetAsComplex128({}) * prev_alpha;
  config.set_alpha_real(new_alpha.real());
  ...
```

The fold is guarded on `beta == 0` (protecting the C matrix) but **not on the epilogue**. Rewrite order for this pattern: `HandleAdd` first fuses the vector bias as `epilogue = BIAS` (beta still 0), then `HandleMultiply` folds the scalar into `alpha` — silently dropping the scale from the bias, wrong by exactly `(1 - alpha) * bias`. The outer add then correctly becomes `beta = 1` with `C = point`. Notably, the neighboring folds in the same file *are* epilogue-aware (e.g. the matrix-bias fold explicitly restricts epilogues, and the `beta*C` fold at [L1886](https://github.com/openxla/xla/blob/2fa50644adea1d5a9bcea781bed98ad29141a4ef/xla/backends/gpu/transforms/gemm_rewriter.cc#L1886) reasons about which epilogues commute) — `HandleMultiply` is the one fold missing the check.

**Why only the autotuner path hits it:** in the main pipeline, `GemmFusion` (Triton) claims these dots before `GemmRewriter` sees them. The buggy fold only executes when the gemm-fusion autotuner's cuBLAS fission candidate re-runs `GemmRewriter` on the extracted fusion body (`FissionBackend::GetFissionedAndRewrittenModule` wrapping `GemmRewriter` in [factory_cuda.cc](https://github.com/openxla/xla/blob/2fa50644adea1d5a9bcea781bed98ad29141a4ef/xla/backends/gpu/autotuner/factory_cuda.cc#L60)) and that candidate wins the timing race — which also explains the shape/batch-size dependence and why closure-constant weights (constant-folded before fusion) don't reproduce.

**Proposed fix** — restrict the fold to plain matmuls:

```cpp
if (config.beta() == 0.0 &&
    config.epilogue() == GemmBackendConfig::DEFAULT &&
    existing_gemm->user_count() == 1) {
```

Any additive epilogue (`BIAS`, `BIAS_RELU`, `BIAS_GELU*`) is unsound under this fold, and the nonlinear ones (`RELU`, `GELU*`) don't commute with scaling either (`relu(alpha*x) != alpha*relu(x)` for `alpha < 0`; GELU never commutes). Keeping the multiply as a separate elementwise op costs almost nothing. A more aggressive alternative — also multiplying the bias operand by the scalar — preserves the fusion but touches more surface; the `DEFAULT`-only guard is the minimal safe fix. A regression test would be: HLO with `multiply(cublas-lt-matmul[epilogue=BIAS](..., bias), scalar)` run through `GemmRewriter`, asserting either no fold or a rescaled bias.

**Comment by adam-hartshorne** [comment]
If its an definitely an XLA issue rather than JAX, does this need to be cross posted to XLA github?

**Comment by jonkhler** [comment]
Agreed — the defect and fix are in XLA, so cross-posted with the compiler-level analysis: https://github.com/openxla/xla/issues/46110. Leaving this issue open too since the user-facing impact (silent wrong results under jit defaults) is what JAX users will search for.
